### DISCUSSION TITLE
how to handle unhandled async function exception?

### DISCUSSION THREAD

**Discussion by fsz1987** [ClearFoundry/ClearScript#407] [General]
I'm using Clearscript v8 .

by default, if an async function throws an error and the caller did not await it.
then the exception is omitted like nothing happened.

example: 
```javascript
//index.js
async function do_some_error (){
    throw new Error ("error!");
}

do_some_error ();

//... some other stuff

```

is there any way to setup a global hook to handle unhandled async function exception/Promise rejection ? 

I hope I can do some fallback process.
for example, print error message and stack trace to the console.

**Comment by ClearScriptLib** [comment]
Hi @fsz1987,

>is there any way to setup a global hook to handle unhandled async function exception/Promise rejection ?

No, but it's something we could look into. What would this feature look like, ideally? For example, are you envisioning a script event, a managed callback, or something else?

**Reply by fsz1987** [reply]
I hope there could be a managed callback or event.
as a host program, if I need handle this in script. I can expose it to the script easily.

**Reply by ClearScriptLib** [reply]
Hi again @fsz1987,

>I hope there could be a managed callback or event.

We've taken a quick look, and it appears doable.

However, unlike .NET's [`UnobservedTaskException`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler.unobservedtaskexception) event, V8's callback is synchronous with promise rejection. That means that simply calling [`Promise.reject`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject) would trigger the callback, and it would be a relatively expensive one due to the transition to managed code.

Would that still be useful to you? If so, V8 offers additional callbacks related to promises:
- Handler added after rejection
- Settled promise resolved
- Settled promise rejected

Would any of those be interesting to you?

Thanks!
